music/sndblit.1
.TH SNDBLIT 1
.SH NAME
sndblit \- sound blitter for making electronic music
.SH SYNOPSIS
.B sndblit
.RB [ \-d ]\ [ \-o\ outputfile ]\ [ \-s\ samplingfrequency ]
.SH DESCRIPTION
.B sndblit
accepts blip descriptors and outputs a RIFF WAVE
.I (\.wav)
file. A "blip" is an instantaneous
snapshot of an audio spectrum. Like the persistence of vision that
makes movies work, the human hearing system perceives continuous sound
when you hear a sound generated by updating an audio spectrum about 30
times per second.
.B sndblit
exploits this persistence of hearing.
.PP
The input to
.B sndblit
is a blip descriptor file. The first line is a floating pointer
number, the blip duration, and an int, the total number of blips in
the file. This is followed by one descriptor for each blip.
.PP
Each blip descriptor starts with one line, the number of sinusoids in
this blip. For each sinusoid there is a line with five floats, the
frequency, left channel cosine component, left channel sine component,
right channel cosine component, and right channel sine component.
.SH OPTIONS
.TP
.B \-i, \-\-input\-file
Specifies an input file. If not provided, blip descriptors are read
from standard input.
.TP
.B \-o, \-\-output\-file
Specifies an output file. If not provided, the default output file is
.I sndblit_output.wav.
.TP
.B \-v, \-\-verbose
Run in verbose mode, printing information about the sinusoids within
each blip.
.TP
.B \-s, \-\-sampling\-frequency
Specifies a new sampling frequency. The default sampling frequency
is 44100 Hertz, audio CD quality.
.TP
.B \-h, \-?, \-\-help
Print helpful(?) usage information.
.SH AUTHOR
Will Ware <ww...@alum.mit.edu> has placed this program in the public
domain.


tinyscheme1.39/CHANGES
     Change Log
     ----------
     Version 1.39
          Drew Yao fixed buffer overflow problems in mk_sharp_const.
     Version 1.38
          Interim release until the rewrite, mostly incorporating modifications from 
          Kevin Cozens. Small addition for Cygwin in the makefile, and modifications
          by Andrew Guenther for Apple platforms.
     Version 1.37
          Joe Buehler submitted reserve_cells.
     Version 1.36
	  Joe Buehler fixed a patch in the allocator.
	  Alexander Shendi moved the comment handling in the scanner, which 
	  fixed an obscure bug for which Mike E had provided a patch as well.
	  Kevin Cozens has submitted some fixes and modifications which have not
	  been incorporated yet in their entirety.
     Version 1.35
	  Todd Showalter discovered that the number of free cells reported 
	  after GC was incorrect, which could also cause unnecessary allocations.
	 Version 1.34
	  Long missing version. Lots of bugfixes have accumulated in my email, so
	  I had to start using them. In this version, Keenan Pepper has submitted
	  a bugfix for the string comparison library procedure, Wouter Boeke 
	  modified some code that was casting to the wrong type and crashed on
	  some machines, "SheppardCo" submitted a replacement "modulo" code and
	  Scott Fenton submitted lots of corrections that shut up some compiler
	  warnings. Brian Maher submitted instructions on how to build on OS-X.
	  I have to dig deeper into my mailbox and find earlier emails, too.
     Version 1.33
	  Charles Hayden fixed a nasty GC bug of the new stack frame, while in
	  the process of porting TinyScheme to C++. He also submitted other 
	  changes, and other people also had comments or requests, but the GC
	  bug was so important that this version is put through the door to 
	  correct it.
     Version 1.32
	  Stephen Gildea put some quality time on TinyScheme again, and made
	  a whole lot of changes to the interpreter that made it noticeably 
	  faster.
     Version 1.31
          Patches to the hastily-done version 1.30. Stephen Gildea fixed
	  some things done wrongly, and Richard Russo fixed the makefile
	  for building on Windows. Property lists (heritage from MiniScheme)
	  are now optional and have dissappeared from the interface. They
	  should be considered as deprecated.
     Version 1.30
	  After many months, I followed Preston Bannister's advice of
	  using macros and a single source text to keep the enums and the
	  dispatch table in sync, and I used his contributed "opdefines.h".
	  Timothy Downs contributed a helpful function, "scheme_call".
	  Stephen Gildea contributed new versions of the makefile and 
	  practically all other sources. He created a built-in STRING-APPEND,
	  and fixed a lot of other bugs.
	  Ruhi Bloodworth reported fixes necessary for OS X and a small
	  bug in dynload.c.
     Version 1.29
	  The previous version contained a lot of corrections, but there
	  were a lot more that still wait on a sheet of paper lost in a
	  carton someplace after my house move... Manuel Heras-Gilsanz
	  noticed this and resent his own contribution, which relies on
	  another bugfix that v.1.28 was missing: a problem with string
	  output, that this version fixes. I hope other people will take
	  the time to resend their contributions, if they didn't make it
	  to v.1.28.
     Version 1.28
	  Many people have contacted me with bugfixes or remarks in
	  the three months I was inactive. A lot of them spotted that 
	  scheme_deinit crashed while reporting gc results. They suggested
	  that sc->outport be set to NIL in scheme_deinit, which I did.
	  Dennis Taylor remarked that OP_VALUEPRINT reset sc->value instead
	  of preserving it. He submitted a modification which I adopted 
	  partially. David Hovemeyer sent me many little changes, that you
	  will find in version 1.28, and Partice Stoessel modified the 
	  float reader to conform to R5RS.
     Version 1.27
          Version 1.27 is the successor of 1.25. Bug fixes only, but I had to
          release them so that everybody can profit. 'Backchar' tried to write
          back to the string, which obviously didn't work for const strings.
          'Substring' didn't check for crossed start and end indices. Defines
          changed to restore the ability to compile under MSVC.
     Version 1.26
          Version 1.26 was never released. I changed a lot of things, in fact
          too much, even the garbage collector, and hell broke loose. I'll
          try a more gradual approach next time.
     Version 1.25
          Types have been homogenized to be able to accomodate a different
	  representation. Plus, promises are no longer closures. Unfortunately,
	  I discovered that continuations and force/delay do not pass the SCM
	  test (and never did)... However, on the bright side, what little
	  modifications I did had a large impact on the footprint: 
	  USE_NO_FEATURES now produces an object file of 63960 bytes on Linux!
     Version 1.24
	  SCM tests now pass again after change in atom2str.
     Version 1.23
          Finally I managed to mess it up with my version control. Version
	  1.22 actually lacked some of the things I have been fixing in the
	  meantime. This should be considered as a complete replacement for
	  1.22.
     Version 1.22
          The new ports had a bug in LOAD. MK_CLOSURE is introduced.
	  Shawn Wagner inquired about string->number and number->string.
	  I added string->atom and atom->string and defined the number
	  functions from them. Doing that, I fixed WRITE applied to symbols
	  (it didn't quote them). Unfortunately, minimum build is now
	  slightly larger than 64k... I postpone action because Jason's idea
	  might solve it elegantly.
     Version 1.21
          Jason Felice submitted a radically different datatype representation

music/sndblit.c
/***************************************************************
NAME
       sndblit - sound blitter for making electronic music

SYNOPSIS
       sndblit [-d] [-o outputfile] [-s samplingfrequency]

DESCRIPTION
       sndblit  accepts  blip descriptors and outputs a RIFF WAVE
       (.wav) file. A "blip" is an instantaneous snapshot  of  an
       audio  spectrum. Like the persistence of vision that makes
       movies work, the human hearing system perceives continuous
       sound when you hear a sound generated by updating an audio
       spectrum about 30 times per second.  sndblit exploits this
       persistence of hearing.

       The  input to sndblit is a blip descriptor file. The first
       line is a floating pointer number, the blip duration,  and
       an  int,  the  total  number of blips in the file. This is
       followed by one descriptor for each blip.

       Each blip descriptor starts with one line, the  number  of
       sinusoids  in this blip. For each sinusoid there is a line
       with five floats, the frequency, left channel cosine  com­
       ponent,  left channel sine component, right channel cosine
       component, and right channel sine component.

OPTIONS
       -i, --input-file
              Specifies an input  file.  If  not  provided,  blip
              descriptors are read from standard input.

       -o, --output-file
              Specifies  an  output  file.  If  not provided, the
              default output file is sndblit_output.wav.

       -v, --verbose
              Run in verbose mode, printing information about the
              sinusoids within each blip.

       -s, --sampling-frequency
              Specifies  a  new  sampling  frequency. The default
              sampling frequency is 44100 Hertz, audio  CD  qual­
              ity.

       -h, -?, --help
              Print helpful(?) usage information.

AUTHOR
       Will  Ware <ww...@alum.mit.edu> has placed this program in
       the public domain.
**********************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>
#include <string.h>
#include <math.h>

#define ASSERT(c) if (!(c)) \
  { fprintf(stderr,"Assertion failed at line %d:\n%s\n", __LINE__, #c); exit(1); }

struct sinusoid {
	double frequency;
	int left_cosine, right_cosine, left_sine, right_sine;
};

struct blip {
	int num_sinusoids;
	struct sinusoid waves[1];	/* growable via malloc */
};

#ifndef M_PI
#define M_PI 3.141592653589793238462
#endif

#define CYCLE 8192		/* this should be a power of 2 */
#define FRACTION 0x1000		/* used for fixed-point math */

static int verbose_flag = 0;

static FILE *infile;
static char outfile_name[80];
static int num_blips, blipsize, max_value;
static int *leftbuf, *rightbuf;
static double blip_duration, sampling_frequency = 44100.0, t, dt;
static struct blip *current_blip = NULL;
static struct blip *previous_blip = NULL;
static int cos_table[CYCLE], sin_table[CYCLE];

#define GET1(fmt,x)          getline(1, fmt, x, NULL, NULL, NULL, NULL)
#define GET2(fmt,x,y)        getline(2, fmt, x, y, NULL, NULL, NULL)
#define GET3(fmt,x,y,z)      getline(3, fmt, x, y, z, NULL, NULL)
#define GET4(fmt,x,y,z,u)    getline(4, fmt, x, y, z, u, NULL)
#define GET5(fmt,x,y,z,u,v)  getline(5, fmt, x, y, z, u, v)

static void getline(int expected, char *format,
		    void *var1, void *var2, void *var3,
		    void *var4, void *var5)
{
	static int input_line_number = 1;
	static char input_line[2001];
	if (fgets(input_line, 2000, infile) == NULL) {
		fprintf(stderr, "Premature end of input at line %d\n",
			input_line_number);
		exit(1);
	}
	input_line[strlen(input_line) - 1] = '\0';
	if (verbose_flag)
		printf("Line %d: \"%s\"\n", input_line_number, input_line);
	if (sscanf(input_line, format, var1, var2, var3, var4, var5)
	    != expected) {
		fprintf(stderr, "Garbage in input at line %d\n",
			input_line_number);
		exit(1);
	}
	input_line_number++;
}

/* Generate stereo .WAV files with 16-bit data. */
#define NCHANNELS 2
#define SAMPWIDTH 2
#define WAVE_FORMAT_PCM 1

static void put2(int x, FILE * outf)
{
	putc(x & 0xFF, outf);
	putc((x >> 8) & 0xFF, outf);
}

static void put4(int x, FILE * outf)
{
	putc(x & 0xFF, outf);
	putc((x >> 8) & 0xFF, outf);
	putc((x >> 16) & 0xFF, outf);
	putc((x >> 24) & 0xFF, outf);
}

static void wave_file_header(FILE * outf)
{
	int datasize;
	datasize = num_blips * blipsize * NCHANNELS * SAMPWIDTH;
	fprintf(outf, "RIFF");
	put4(datasize + 36, outf);
	fprintf(outf, "WAVEfmt ");
	put4(16, outf);
	put2(WAVE_FORMAT_PCM, outf);
	put2(NCHANNELS, outf);
	put4((int) sampling_frequency, outf);
	put4((int) (sampling_frequency * NCHANNELS * SAMPWIDTH), outf);
	put2(NCHANNELS * SAMPWIDTH, outf);
	put2(8 * SAMPWIDTH, outf);
	fprintf(outf, "data");
	put4(datasize, outf);
}

static int compare_frequencies(const void *arg1, const void *arg2)
{
	struct sinusoid *s1, *s2;
	s1 = (struct sinusoid *) arg1;
	s2 = (struct sinusoid *) arg2;
	if (s1->frequency == s2->frequency)
		return 0;
	if (s1->frequency < s2->frequency)
		return -1;
	return 1;
}


#define TESTMALLOC(x) \
  if ((x) == NULL) { \
    fprintf(stderr, "Line %d: out of memory\n", __LINE__); \
    exit(1); }

static void read_in_blip(void)
{
	int j, num_sinusoids;
	double lc, ls, rc, rs;
	GET1("%d", &num_sinusoids);
	if (verbose_flag)
		printf("reading blip, %d sinusoids\n", num_sinusoids);
	current_blip = malloc(sizeof(struct blip) +
			      (num_sinusoids -
			       1) * sizeof(struct sinusoid));
	TESTMALLOC(current_blip);
	current_blip->num_sinusoids = num_sinusoids;
	for (j = 0; j < num_sinusoids; j++) {
		GET5("%lf %lf %lf %lf %lf",
		     &(current_blip->waves[j].frequency),
		     &lc, &ls, &rc, &rs);
		current_blip->waves[j].left_cosine = (int) (FRACTION * lc);
		current_blip->waves[j].right_cosine =
		    (int) (FRACTION * rc);
		current_blip->waves[j].left_sine = (int) (FRACTION * ls);
		current_blip->waves[j].right_sine = (int) (FRACTION * rs);
		if (verbose_flag)
			printf("f=%f lc=%d rc=%d ls=%d rs=%d\n",
			       current_blip->waves[j].frequency,
			       current_blip->waves[j].left_cosine,
			       current_blip->waves[j].right_cosine,
			       current_blip->waves[j].left_sine,
			       current_blip->waves[j].right_sine);
	}
	qsort(current_blip->waves, num_sinusoids,
	      sizeof(struct sinusoid), compare_frequencies);
}

static void render_sinusoid(int *buf, double f,
			    int cos_start, int cos_finish,
			    int sin_start, int sin_finish)
{
	int i;
	int c0, c1, dc0, dc1;
	int s0, s1, ds0, ds1;
	int ph1, ph0, dph1, dph0;

	{
		double cos_delta =
		    ((double) (cos_finish - cos_start)) / blipsize;
		double sin_delta =
		    ((double) (sin_finish - sin_start)) / blipsize;
		c1 = cos_start;
		c0 = 0;
		dc1 = (int) cos_delta;
		dc0 = (int) (FRACTION * (cos_delta - dc1));
		s1 = sin_start;
		s0 = 0;
		ds1 = (int) sin_delta;
		ds0 = (int) (FRACTION * (sin_delta - ds1));
	}

	{
		double phase, phase_delta;
		phase = CYCLE * f * t;
		phase_delta = CYCLE * f * blip_duration / blipsize;
		ph1 = (int) phase;
		ph0 = (int) (FRACTION * (phase - ph1));
		dph1 = (int) phase_delta;
		dph0 = (int) (FRACTION * (phase_delta - dph1));
	}

	for (i = 0; i < blipsize; i++) {
		buf[i] +=
		    c1 * cos_table[ph1 & (CYCLE - 1)] +
		    s1 * sin_table[ph1 & (CYCLE - 1)];

		c0 += dc0;
		c1 += dc1;
		while (c0 >= FRACTION) {
			c1++;
			c0 -= FRACTION;
		}
		while (c0 < -FRACTION) {
			c1--;
			c0 += FRACTION;
		}

		s0 += ds0;
		s1 += ds1;
		while (s0 >= FRACTION) {
			s1++;
			s0 -= FRACTION;
		}
		while (s0 < -FRACTION) {
			s1--;
			s0 += FRACTION;
		}

		ph0 += dph0;
		ph1 += dph1;
		while (ph0 >= FRACTION) {
			ph1++;
			ph0 -= FRACTION;
		}
		while (ph0 < -FRACTION) {
			ph1--;
			ph0 += FRACTION;
		}
	}
}

static void merge_with_previous_blip(void)
{
	int i, j, b_abs;
	i = j = 0;
	while (1) {
		if (i < current_blip->num_sinusoids &&
		    j < previous_blip->num_sinusoids &&
		    current_blip->waves[i].frequency ==
		    previous_blip->waves[j].frequency) {
			/* this is the same sinusoid spanning two blips */
			if (verbose_flag)
				printf
				    ("continue existing sinusoid, f=%f\n",
				     current_blip->waves[i].frequency);
			render_sinusoid(leftbuf,
					current_blip->waves[i].frequency,
					previous_blip->waves[j].
					left_cosine,
					current_blip->waves[i].left_cosine,
					previous_blip->waves[j].left_sine,
					current_blip->waves[i].left_sine);
			render_sinusoid(rightbuf,
					current_blip->waves[i].frequency,
					previous_blip->waves[j].
					right_cosine,
					current_blip->waves[i].
					right_cosine,
					previous_blip->waves[j].right_sine,
					current_blip->waves[i].right_sine);
			i++;
			j++;
		} else if (i < current_blip->num_sinusoids &&
			   (j == previous_blip->num_sinusoids ||
			    current_blip->waves[i].frequency <
			    previous_blip->waves[j].frequency)) {
			/* this is the beginning of a new sinusoid */
			if (verbose_flag)
				printf("begin new sinusoid, f=%f\n",
				       current_blip->waves[i].frequency);
			render_sinusoid(leftbuf,
					current_blip->waves[i].frequency,
					0,
					current_blip->waves[i].left_cosine,
					0,
					current_blip->waves[i].left_sine);
			render_sinusoid(rightbuf,
					current_blip->waves[i].frequency,
					0,
					current_blip->waves[i].
					right_cosine, 0,
					current_blip->waves[i].right_sine);
			i++;
		} else if (j < previous_blip->num_sinusoids &&
			   (i == current_blip->num_sinusoids ||
			    previous_blip->waves[j].frequency <
			    current_blip->waves[i].frequency)) {
			/* this is the end of an old sinusoid */
			if (verbose_flag)
				printf("end old sinusoid, f=%f\n",
				       previous_blip->waves[j].frequency);
			render_sinusoid(leftbuf,
					previous_blip->waves[j].frequency,
					previous_blip->waves[j].
					left_cosine, 0,
					previous_blip->waves[j].left_sine,
					0);
			render_sinusoid(rightbuf,
					previous_blip->waves[j].frequency,
					previous_blip->waves[j].
					right_cosine, 0,
					previous_blip->waves[j].right_sine,
					0);
			j++;
		} else
			break;
	}
	for (i = b_abs = 0; i < blipsize; i++) {
		if (leftbuf[i] < -b_abs)
			b_abs = -leftbuf[i];
		else if (leftbuf[i] > b_abs)
			b_abs = leftbuf[i];
		if (rightbuf[i] < -b_abs)
			b_abs = -rightbuf[i];
		else if (rightbuf[i] > b_abs)
			b_abs = rightbuf[i];
	}
	if (max_value < b_abs)
		max_value = b_abs;
	if (verbose_flag)
		printf("max value = %08x\n", max_value);
}

static void help(void)
{
	printf("NAME\n"
	       "       sndblit - sound blitter for making electronic music\n"
	       "\n"
	       "SYNOPSIS\n"
	       "       sndblit [-d] [-o outputfile] [-s samplingfrequency]\n"
	       "\n"
	       "DESCRIPTION\n"
	       "       sndblit  accepts  blip descriptors and outputs a RIFF WAVE\n"
	       "       (.wav) file. A \"blip\" is an instantaneous snapshot  of  an\n"
	       "       audio  spectrum. Like the persistence of vision that makes\n"
	       "       movies work, the human hearing system perceives continuous\n"
	       "       sound when you hear a sound generated by updating an audio\n"
	       "       spectrum about 30 times per second.  sndblit exploits this\n"
	       "       persistence of hearing.\n"
	       "\n"
	       "       The  input to sndblit is a blip descriptor file. The first\n"
	       "       line is a floating pointer number, the blip duration,  and\n"
	       "       an  int,  the  total  number of blips in the file. This is\n"
	       "       followed by one descriptor for each blip.\n"
	       "\n"
	       "       Each blip descriptor starts with one line, the  number  of\n"
	       "       sinusoids  in this blip. For each sinusoid there is a line\n"
	       "       with five floats, the frequency, left channel cosine  com­\n"
	       "       ponent,  left channel sine component, right channel cosine\n"
	       "       component, and right channel sine component.\n"
	       "\n"
	       "OPTIONS\n"
	       "       -i, --input-file\n"
	       "              Specifies an input  file.  If  not  provided,  blip\n"
	       "              descriptors are read from standard input.\n"
	       "\n"
	       "       -o, --output-file\n"
	       "              Specifies  an  output  file.  If  not provided, the\n"
	       "              default output file is sndblit_output.wav.\n"
	       "\n"
	       "       -v, --verbose\n"
	       "              Run in verbose mode, printing information about the\n"
	       "              sinusoids within each blip.\n"
	       "\n"
	       "       -s, --sampling-frequency\n"
	       "              Specifies  a  new  sampling  frequency. The default\n"
	       "              sampling frequency is 44100 Hertz, audio  CD  qual­\n"
	       "              ity.\n"
	       "\n"
	       "       -h, -?, --help\n"
	       "              Print helpful(?) usage information.\n"
	       "\n"
	       "AUTHOR\n"
	       "       Will  Ware <ww...@alum.mit.edu> has placed this program in\n"
	       "       the public domain.\n");
}


static const struct option long_options[] = {
	{"verbose", no_argument, NULL, 'v'},
	{"input-file", required_argument, NULL, 'i'},
	{"output-file", required_argument, NULL, 'o'},
	{"sampling-frequency", required_argument, NULL, 's'},
	{"help", no_argument, NULL, 'h'},
	{NULL, 0, NULL, 0}
};

int main(int argc, char *argv[])
{
	FILE *outf, *leftf, *rightf;
	int c, i, j, k, mult;
	unsigned char *outbuf;
	static struct blip *zero_blip;

	infile = stdin;
	strcpy(outfile_name, "sndblit_output.wav");
	while ((c = getopt_long(argc, argv, "s:i:o:vh?",
				long_options, (int *) 0)) != EOF)
		switch (c) {
		case 's':
			sampling_frequency = atof(optarg);
			break;
		case 'i':
			infile = fopen(optarg, "r");
			if (infile == NULL) {
				fprintf(stderr,
					"cannot open \"%s\" for reading\n",
					optarg);
				return 1;
			}
			break;
		case 'o':
			strcpy(outfile_name, optarg);
			break;
		case 'v':
			verbose_flag = 1;
			break;
		default:
			help();
			return 0;
		}
	for (i = 0; i < CYCLE; i++) {
		double angle = (2.0 * M_PI / CYCLE) * i;
		cos_table[i] = (int) (FRACTION * cos(angle));
		sin_table[i] = (int) (FRACTION * sin(angle));
	}

	GET2("%lf %d", &blip_duration, &num_blips);
	num_blips += 2;		/* we'll add extra blips at beginning and end */
	dt = 1. / sampling_frequency;
	blipsize = (int) (blip_duration / dt);
	/* At the cost of a slight inaccuracy in tempo but with a vast
	 * simplification of the code, coerce blip_duration to be an exact
	 * multiple of dt.
	 */
	blip_duration = dt * blipsize;
	leftbuf = malloc(blipsize * sizeof(int));
	TESTMALLOC(leftbuf);
	rightbuf = malloc(blipsize * sizeof(int));
	TESTMALLOC(rightbuf);

	/* make a blip with zero amplitudes */
	zero_blip = malloc(sizeof(struct blip) - sizeof(struct sinusoid));
	TESTMALLOC(zero_blip);
	zero_blip->num_sinusoids = 0;
	current_blip = zero_blip;

	leftf = fopen("left.tmp", "wb");
	if (leftf == NULL) {
		fprintf(stderr, "can't open left.tmp for writing\n");
		return 1;
	}
	rightf = fopen("right.tmp", "wb");
	if (rightf == NULL) {
		fprintf(stderr, "can't open right.tmp for reading\n");
		return 1;
	}
	max_value = 0;

	/* step thru all the blips in the input */
	for (i = 0, t = 0.0; i < num_blips; i++, t += blipsize * dt) {
		if (verbose_flag)
			printf("starting loop for blip %d\n", i - 1);
		memset(leftbuf, 0, blipsize * sizeof(int));
		memset(rightbuf, 0, blipsize * sizeof(int));
		if (previous_blip != NULL && previous_blip != zero_blip)
			free(previous_blip);
		previous_blip = current_blip;
		if (i == 0 || i == num_blips - 1)
			/* at the beginning, ramp up from the zero blip */
			/* at the end, ramp down to the zero blip */
			current_blip = zero_blip;
		else
			read_in_blip();
		merge_with_previous_blip();
		fwrite(leftbuf, sizeof(int), blipsize, leftf);
		fwrite(rightbuf, sizeof(int), blipsize, rightf);
		if (verbose_flag)
			printf("finishing loop for blip %d\n", i - 1);
	}
	fclose(leftf);
	fclose(rightf);
	fclose(infile);
	if (previous_blip != NULL && previous_blip != zero_blip)
		free(previous_blip);
	free(zero_blip);
	mult = (int) ((32700.0 * 65536.0) / max_value);
	if (verbose_flag)
		printf("max_value = %d, mult = %d\n", max_value, mult);

	outf = fopen(outfile_name, "wb");
	if (outf == NULL) {
		fprintf(stderr, "can't open \"%s\" for writing\n",
			outfile_name);
		return 1;
	}
	wave_file_header(outf);
	outbuf = malloc(4 * blipsize);
	TESTMALLOC(outbuf);
	leftf = fopen("left.tmp", "rb");
	if (leftf == NULL) {
		fprintf(stderr, "can't open left.tmp for reading\n");
		return 1;
	}
	rightf = fopen("right.tmp", "rb");
	if (rightf == NULL) {
		fprintf(stderr, "can't open right.tmp for reading\n");
		return 1;
	}
	for (i = 0; i < num_blips; i++) {
		ASSERT(fread(leftbuf, sizeof(int), blipsize, leftf) ==
		       blipsize);
		ASSERT(fread(rightbuf, sizeof(int), blipsize, rightf) ==
		       blipsize);
		for (j = k = 0; j < blipsize; j++) {
			int x, y;
			x = (mult * leftbuf[j]) >> 16;
			y = (mult * rightbuf[j]) >> 16;
			outbuf[k++] = x & 0xFF;
			outbuf[k++] = (x >> 8) & 0xFF;
			outbuf[k++] = y & 0xFF;
			outbuf[k++] = (y >> 8) & 0xFF;
		}
		ASSERT(k == 4 * blipsize);
		ASSERT(fwrite(outbuf, 4, blipsize, outf) == blipsize);
	}
	ASSERT(fclose(leftf) == 0);
	ASSERT(fclose(rightf) == 0);
	ASSERT(fclose(outf) == 0);
	free(leftbuf);
	free(rightbuf);
	free(outbuf);
	system("rm -f left.tmp right.tmp");
	return 0;
}
